src/covers/discogscoverprovider.cpp

```diff
@@ -0,0 +1,373 @@
+    : CoverProvider("Discogs", false, parent),
+      network_(new NetworkAccessManager(this)) {}
+
+bool DiscogsCoverProvider::StartSearch(const QString& artist,
+                                       const QString& album, int s_id) {
+  DiscogsCoverSearchContext* s_ctx = new DiscogsCoverSearchContext;
+  if (s_ctx == nullptr) return false;
+  s_ctx->id = s_id;
+  s_ctx->artist = artist;
+  s_ctx->album = album;
+  s_ctx->r_count = 0;
+  requests_search_.insert(s_id, s_ctx);
+  SendSearchRequest(s_ctx);
+
+  return true;
+}
+
+void DiscogsCoverProvider::CancelSearch(int id) {
+  delete requests_search_.take(id);
+}
+
+bool DiscogsCoverProvider::StartRelease(DiscogsCoverSearchContext* s_ctx,
+                                        int r_id, QString resource_url) {
+  DiscogsCoverReleaseContext* r_ctx = new DiscogsCoverReleaseContext;
+  if (r_ctx == nullptr) return false;
+
+  s_ctx->r_count++;
+
+  r_ctx->id = r_id;
+  r_ctx->resource_url = resource_url;
+
+  r_ctx->s_id = s_ctx->id;
+
+  requests_release_.insert(r_id, r_ctx);
+  SendReleaseRequest(s_ctx, r_ctx);
+
+  return true;
+}
+
+void DiscogsCoverProvider::SendSearchRequest(DiscogsCoverSearchContext* s_ctx) {
+  typedef QPair<QString, QString> Arg;
+  typedef QList<Arg> ArgList;
+
+  typedef QPair<QByteArray, QByteArray> EncodedArg;
+  typedef QList<EncodedArg> EncodedArgList;
+
+  ArgList args =
+      ArgList() << Arg("key", QByteArray::fromBase64(kAccessKeyB64))
+                << Arg("secret", QByteArray::fromBase64(kSecretKeyB64));
+
+  args.append(Arg("type", "release"));
+  if (!s_ctx->artist.isEmpty()) {
+    args.append(Arg("artist", s_ctx->artist.toLower()));
+  }
+  if (!s_ctx->album.isEmpty()) {
+    args.append(Arg("release_title", s_ctx->album.toLower()));
+  }
+
+  QUrlQuery url_query;
+  QUrl url(kUrlSearch);
+  QStringList query_items;
+
+  // Encode the arguments
+  for (const Arg& arg : args) {
+    EncodedArg encoded_arg(QUrl::toPercentEncoding(arg.first),
+                           QUrl::toPercentEncoding(arg.second));
+    query_items << QString(encoded_arg.first + "=" + encoded_arg.second);
+    url_query.addQueryItem(encoded_arg.first, encoded_arg.second);
+  }
+
+  // Sign the request
+  const QByteArray data_to_sign =
+      QString("GET\n%1\n%2\n%3")
+          .arg(url.host(), url.path(), query_items.join("&"))
+          .toLatin1();
+  const QByteArray signature(Utilities::HmacSha256(
+      QByteArray::fromBase64(kSecretKeyB64), data_to_sign));
+
+  // Add the signature to the request
+  url_query.addQueryItem("Signature",
+                         QUrl::toPercentEncoding(signature.toBase64()));
+
+  url.setQuery(url_query);
+  QNetworkReply* reply = network_->get(QNetworkRequest(url));
+
+  NewClosure(reply, SIGNAL(error(QNetworkReply::NetworkError)), this,
+             SLOT(SearchRequestError(QNetworkReply::NetworkError,
+                                     QNetworkReply*, int)),
+             reply, s_ctx->id);
```

Comment:
I don't think you even need all these ids and the two `QHash`. You can just pass the context objects around instead via `NewClosure`.